src/outputs/history.cpp

```diff
@@ -42,7 +44,108 @@ namespace parthenon {
 //  \brief Writes a history file
 
 void HistoryOutput::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
-  throw std::runtime_error(std::string(__func__) + " is not implemented");
+  std::vector<std::string> all_labels = {};
+  std::vector<Real> all_results = {};
+
+  // Loop over all packages of the application
+  for (const auto &pkg : pm->packages.AllPackages()) {
+    // Check if the package has enrolled functions which are stored in the
+    // Params under the `hist_param_key` name.
+    const auto &params = pkg.second->AllParams();
+    if (!params.hasKey(hist_param_key)) {
+      continue;
+    }
+    auto hist_vars = params.Get<HstVar_list>(hist_param_key);
+
+    for (auto &hist_var : hist_vars) {
+      // Get "base" MeshData, which always exists but may not be populated yet
+      auto &md_base = pm->mesh_data.Get();
+      // Populated with all blocks
+      if (md_base->NumBlocks() == 0) {
+        md_base->Set(pm->block_list, "base");
+      } else if (md_base->NumBlocks() != pm->block_list.size()) {
+        PARTHENON_WARN(
+            "Resetting \"base\" MeshData to contain all blocks. This indicates that "
+            "the \"base\" MeshData container has been modified elsewhere. Double check "
+            "that the modification was intentional and is compatible with this reset.")
+        md_base->Set(pm->block_list, "base");
+      }
+      auto result = hist_var.hst_fun(md_base.get());
+#ifdef MPI_PARALLEL
+      // need fence so that the result is ready prior to the MPI call
+      Kokkos::fence();
+      // apply separate chosen operations to each user-defined history output
+      MPI_Op usr_op;
+      switch (hist_var.hst_op) {
+      case UserHistoryOperation::sum:
+        usr_op = MPI_SUM;
+        break;
+      case UserHistoryOperation::max:
+        usr_op = MPI_MAX;
+        break;
+      case UserHistoryOperation::min:
+        usr_op = MPI_MIN;
+        break;
+      }
+      if (Globals::my_rank == 0) {
+        PARTHENON_MPI_CHECK(MPI_Reduce(MPI_IN_PLACE, &result, 1, MPI_PARTHENON_REAL,
+                                       usr_op, 0, MPI_COMM_WORLD));
+      } else {
+        PARTHENON_MPI_CHECK(MPI_Reduce(&result, &result, 1, MPI_PARTHENON_REAL, usr_op, 0,
+                                       MPI_COMM_WORLD));
+      }
+#endif
+
+      all_results.emplace_back(result);
+      all_labels.emplace_back(hist_var.label);
```

Comment:
Call reserve before the loop since you alread know its size.

Reply:
The history output collects results from all packages. Thus, a single `reserve` would mean to query all packages first (adding another for loop over packages) or call a reserve for each package.
I don't expect this to be a performance issues (as the history output is usually only called everyone once in a while and the number of variables is typically small, say < 10-20). Thus, there should be no overhead for adding a `string` and one `double` in the loop.
So I prefer to keep the code here as is as it's logically the simplest.

Reply:
Sounds good!